app/src/main/java/com/zulip/android/networking/HTTPRequest.java

```diff
@@ -1,141 +1,115 @@
-        AndroidHttpClient httpclient = AndroidHttpClient.newInstance(app
-                .getUserAgent());
-
-        String responseString = null;
-        try {
-            String url = app.getServerURI() + path;
-            nameValuePairs.add(new BasicNameValuePair("client", "Android"));
-
-            if (method.equals("POST")) {
-                request = new HttpPost(url);
-            } else if (method.equals("PUT")) {
-                request = new HttpPut(url);
-            } else if (method.equals("DELETE")) {
-                request = new HttpDeleteWithReq(url);
-            } else if (method.equals("GET")) {
-                request = new HttpGet(url + "?"
-                        + URLEncodedUtils.format(nameValuePairs, "utf-8"));
+    private String generateURL(String url, HashMap<String, String> properties) {
+        if (properties.isEmpty()) return url;
+        String encodedUrl = url + "?";
+        for (Map.Entry<String, String> map : properties.entrySet()) {
+            try {
+                encodedUrl += URLEncoder.encode(map.getKey(), ("utf-8")) + "=" + URLEncoder.encode(map.getValue(), "utf-8") + "&";
+            } catch (UnsupportedEncodingException e) {
+                ZLog.logException(e);
             }
-
-            if (!method.equals("GET")) {
-                ((HttpEntityEnclosingRequestBase) request)
-                        .setEntity(new UrlEncodedFormEntity(nameValuePairs,
-                                "UTF-8"));
-            }
-
-            Log.i("HTTP.request", request.getMethod() + " " + request.getURI());
-
-            if (this.app.getApiKey() != null) {
-                String authstr = this.app.getEmail() + ":" + this.app.getApiKey();
-                request.setHeader(
-                        "Authorization",
-                        "Basic "+ Base64.encodeToString(authstr.getBytes(), Base64.NO_WRAP));
-            }
-            // timeout after 60 seconds in ms
-            HttpConnectionParams.setSoTimeout(request.getParams(), 60 * 1000);
-
-            response = httpclient.execute(request);
-
-            StatusLine statusLine = response.getStatusLine();
-
-            ByteArrayOutputStream out = new ByteArrayOutputStream();
-            response.getEntity().writeTo(out);
-            out.close();
-            responseString = out.toString();
-            if (statusLine.getStatusCode() != HttpStatus.SC_OK) {
-                Log.e("HTTP", statusLine.getReasonPhrase());
-                Log.e("HTTP", responseString);
-                throw new HttpResponseException(statusLine.getStatusCode(),
-                        responseString);
-            }
-        } finally {
-            httpclient.close();
         }
-
-        if (responseString != null) {
-            Log.i("HTTP.response", responseString);
-        } else {
-            Log.i("HTTP.response", "<empty>");
-        }
-        return responseString;
+        return encodedUrl.substring(0, encodedUrl.length() - 1);
```

Comment:
there's probably a safer way to removing the last character only if it's an `&`.